src/widgets/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,711 @@
+
+                  const auto proportion = srcItem->GetProportion();
+                  const auto flag = srcItem->GetFlag();
+                  const auto border = srcItem->GetBorder();
+                  
+                  boxSizer->Remove(mSourceIndex);
+                  boxSizer->Insert(
+                     mTargetIndex < mSourceIndex ? mTargetIndex : mTargetIndex - 1,
+                     this,
+                     proportion,
+                     flag,
+                     border
+                  );
+               }
+            }
+         }
+         DragFinished();
+      }
+      
+      void OnMove(wxMouseEvent& evt)
+      {
+         if(!mDragging)
+            return;
+
+         auto parent = GetParent();
+         if(!parent)
+            return;
+
+         wxPoint newPosition = wxGetMousePosition() - mInitialPosition;
+         Move(GetParent()->ScreenToClient(newPosition));
+
+         if(auto boxSizer = dynamic_cast<wxBoxSizer*>(parent->GetSizer()))
+         {
+            if(boxSizer->GetOrientation() == wxVERTICAL)
+            {
+               auto targetIndex = mSourceIndex;
+               
+               //assuming that items are ordered from top to bottom (i > j <=> y(i) > y(j))
+               //compare wxSizerItem position with the current MovableControl position!
+               if(GetPosition().y < boxSizer->GetItem(mSourceIndex)->GetPosition().y)
+               {
+                  //moving up
+                  for(int i = 0; i < mSourceIndex; ++i)
+                  {
+                     const auto item = boxSizer->GetItem(i);
+                     
+                     if(GetRect().GetTop() <= item->GetPosition().y + item->GetSize().y / 2)
+                     {
+                        targetIndex = i;
+                        break;
+                     }
+                  }
+               }
+               else
+               {
+                  //moving down
+                  for(int i = static_cast<int>(boxSizer->GetItemCount()) - 1; i > mSourceIndex; --i)
+                  {
+                     const auto item = boxSizer->GetItem(i);
+                     if(GetRect().GetBottom() >= item->GetPosition().y + item->GetSize().y / 2)
+                     {
+                        targetIndex = i + 1;
+                        break;
+                     }
+                  }
+               }
+
+               if(targetIndex != mTargetIndex)
+               {
+                  mTargetIndex = targetIndex;
+                  PostDragEvent(parent, EVT_MOVABLE_CONTROL_DRAG_POSITION);
+               }
+            }
+         }
+      }
+   };
+
+   wxDEFINE_EVENT(EVT_EFFECT_ITEM_REMOVE, wxCommandEvent);
+
+   //UI control that represents individual effect from the effect list
+   class RealtimeEffectControl : public MovableControl
+   {
+      std::shared_ptr<Track> mTrack;
+      //int mEffectIndex{-1};
+      RealtimeEffectState* mEffectState{nullptr};
```

Comment:
I see now that there is a serious problem in this design.

There should be undo and redo of changes to the effect list.  This is not yet part of this pull request.  I anticipated the need at commit 5da76c6eae0aceb1823be43bcb9009aa20ead755 with some support.

If this window persists through pushes and pops of the undo stack, then storing this pointer is the wrong thing.  It might point into a version of the state that is in undo history but not the present project state, and it might become a dangling pointer if redo states are destroyed after an undo and then other edits.

Should the parent window just rebuild itself entirely whenever the undo stack is pushed or popped?  Maybe that's easiest to do.  Just subscribe to events from UndoManager.

In fact, you don't have a choice.  If you can't predict at all how the composition of the RealtimeEffectList changes with undo or redo, you just have to read it all and rebuild.

Okay, that eliminates the dangling pointer danger.